testing/integration/agent_long_test.go

```diff
@@ -0,0 +1,282 @@
+
+//go:build integration
+
+package integration
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"os"
+	"os/exec"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/sajari/regression"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/elastic/elastic-agent-libs/kibana"
+	"github.com/elastic/elastic-agent/pkg/control/v2/cproto"
+	atesting "github.com/elastic/elastic-agent/pkg/testing"
+	"github.com/elastic/elastic-agent/pkg/testing/define"
+	"github.com/elastic/elastic-agent/pkg/testing/tools"
+	"github.com/elastic/go-sysinfo"
+	"github.com/elastic/go-sysinfo/types"
+)
+
+type ExtendedRunner struct {
+	suite.Suite
+	info         *define.Info
+	agentFixture *atesting.Fixture
+
+	ESHost string
+}
+
+type ComponentMetrics struct {
+	Memory        MemoryMetrics  `mapstructure:"memstats"`
+	Handles       HandlesMetrics `mapstructure:"handles"`
+	UnixTimestamp int64
+}
+
+// TestComponent is used as a key in our map of component metrics
+type TestComponent struct {
+	Binary   string `mapstructure:"binary"`
+	Dataset  string `mapstructure:"dataset"`
+	ID       string `mapstructure:"id"`
+	CompType string `mapstructure:"type"`
+}
+
+type MemoryMetrics struct {
+	GcNext      uint64 `mapstructure:"gc_next"`
+	MemoryAlloc uint64 `mapstructure:"memory_alloc"`
+	MemorySys   uint64 `mapstructure:"memory_sys"`
+	MemoryTotal uint64 `mapstructure:"memory_total"`
+	RSS         uint64 `mapstructure:"rss"`
+}
+
+type HandlesMetrics struct {
+	Open  int           `mapstructure:"open"`
+	Limit HandlesLimits `mapstructure:"limit"`
+}
+
+type HandlesLimits struct {
+	Hard uint   `mapstructure:"hard"`
+	Soft uint64 `mapstructure:"soft"`
+}
+
+// MetricsSystem is used for windows handles metrics
+type MetricsSystem struct {
+	Handles HandlesMetrics `mapstructure:"handles"`
+}
+
+type processWatcher struct {
+	handle     types.Process
+	pid        int
+	name       string
+	regHandles *regression.Regression
+}
+
+func TestLongRunningAgentForLeaks(t *testing.T) {
+	info := define.Require(t, define.Requirements{
+		Group: "fleet",
+		Stack: &define.Stack{},
+		Local: false, // requires Agent installation
+		Sudo:  true,  // requires Agent installation
+		OS: []define.OS{
+			{Type: define.Linux},
+			{Type: define.Windows},
+		},
+	})
+
+	if os.Getenv("TEST_LONG_RUNNING") == "" {
+		t.Skipf("not running extended test unless TEST_LONG_RUNNING is set")
+	}
+
+	suite.Run(t, &ExtendedRunner{info: info})
+}
+
+func (runner *ExtendedRunner) SetupSuite() {
+	// create ~40 1MB files that will be picked up by the `/var/log/httpd/error_log*` pattern
+	cmd := exec.Command("go", "install", "-v", "github.com/mingrammer/flog@latest")
+	out, err := cmd.CombinedOutput()
+	require.NoError(runner.T(), err, "got out: %s", string(out))
+
+	cmd = exec.Command("flog", "-t", "log", "-f", "apache_error", "-o", "/var/log/httpd/error_log", "-b", "50485760", "-p", "1048576")
```

Comment:
since we want a long running test, don't we want to constantly generate log files?  And I think we want to rotate them as well.

Reply:
This will generate multiple files with names like `$FILE.1`, wouldn't that be the same as rotating them, from filebeat's perspective?

Reply:
No, not the same at all.  If this makes 40 files at the start of the test, then filebeat will "see" all 40 files when it starts, process them and then close the file handles.  If we have say 5 files, and then rotate them over time then the harvester has to open new files, and correctly close them over time, so it should only have 5 open handles at any given time.  This would catch things like leaking file handles in the filestream fingerprint option.

Reply:
Also, having spigot write continuously seems to produce file-handle-leak-like behavior, where it's just continuously opening large numbers of files to read from. Unless there's some flag I can set through fleet to tell it to be more aggressive with closing file handles?

Reply:
> Also, having spigot write continuously seems to produce file-handle-leak-like behavior, where it's just continuously opening large numbers of files to read from. Unless there's some flag I can set through fleet to tell it to be more aggressive with closing file handles?

If you are always adding new files, filebeat will close the file handle after [close.on_state_change.inactive](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-filestream.html#filebeat-input-filestream-close-inactive)

But we should probably add output rotation to spigot.  Pretty sure it would be easy to add.  I can look next week.  That will let us make sure that [close.on_state_change.removed](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-filestream.html#filebeat-input-filestream-close-removed) is also working and not leaking file handles.

Reply:
If spigot doesn't have this now, is there something we need to change for this PR?